core/runtime/src/main/java/io/quarkus/runtime/metrics/MetricsFactory.java

```diff
@@ -0,0 +1,141 @@
+package io.quarkus.runtime.metrics;
+
+import java.time.Duration;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.function.ToDoubleFunction;
+
+/**
+ * Extensions can create or register metrics using this factory
+ * independent of the enabled metrics provider
+ */
+public interface MetricsFactory {
+
+    /** A well-known string for MicroProfile metrics provided by the SmallRye Metrics quarkus extension */
+    final String MP_METRICS = "smallrye-metrics";
+
+    /** A well-known string for MicroProfile metrics provided by the SmallRye Metrics quarkus extension */
+    final String MICROMETER = "micrometer";
+
+    /**
+     * @return true if this factory supports the named metrics system. Arbitrary
+     *         strings are allowed. Constants are present for a few.
+     * @see #MICROMETER
+     * @see #MP_METRICS
+     */
+    boolean metricsSystemSupported(String name);
+
+    /** Create a fluid builder for registering counters and gauges */
+    MetricBuilder builder(String name);
+
+    interface MetricBuilder {
+        /**
+         * @param description Description text of the eventual metric (optional).
+         * @return The builder with added description.
+         */
+        MetricBuilder description(String description);
+
+        /**
+         * @param key The tag key.
+         * @param value The tag value.
+         * @return The builder with added tag.
+         */
+        MetricBuilder tag(String key, String value);
+
+        /**
+         * Specify the metric unit (optional)
+         * 
+         * @param unit Base unit of the eventual metric
+         * @return The builder with added base unit.
+         */
+        MetricBuilder unit(String unit);
+
+        /**
+         * Register a counter that retrieves its value from a supplier function
+         * 
+         * @param countFunction Function supplying a monotonically increasing number value
+         */
+        void counter(Supplier<Number> countFunction);
```

Comment:
What if an extension needs to create a counter that will be updated via calls to `inc()` rather than be a supplier or function? I guess we need a `counter()` function here without parameters that just registers  a plain counter

Reply:
my intent was to use supplier or ToDoubleFunction always via the shared API. IMO, it is a best practice for reusable libraries anyway.

Reply:
Well, but it would force some libraries to make significant changes, because they work by calling `inc()` and don't create any state objects for their counters.

Reply:
I am going to go find out how common that is. So far, it isn't very. And yes, this is on me to achieve. The goal of this common facade is to make that transition easier... but the model here is what Hibernate, Agroal, and Vert.x use..